packages/wrangler/src/__tests__/file-registry.test.ts

```diff
@@ -0,0 +1,817 @@
+			}
+		}
+	});
+
+	it("should ensure there is no race condition between file watcher setup and stopRegisterWatcher", async () => {
+		const cb = vi.fn();
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars, unused-imports/no-unused-vars
+		void registry.startRegistryWatcher(cb);
+		// This could be another process stopping the registry watcher
+		await registry.stopRegistryWatcher();
+
+		const workerName = "test-worker";
+		const workerDefinition: WorkerDefinition = {
+			port: 8787,
+			protocol: "http",
+			host: "localhost",
+			mode: "local",
+			durableObjects: [],
+		};
+
+		// Register the worker
+		await registry.registerWorker(workerName, workerDefinition);
+
+		// Verify that the watcher was stopped correctly
+		expect(cb).toHaveBeenCalledTimes(0);
+	});
+
+	// TODO: the test fails because we have not awaited the completion of the registration
+	it.skip("should maintain consistency with concurrent unregister and register operations", async () => {
+		const workerName = "test-worker";
+		const workerDefinition: WorkerDefinition = {
+			port: 8787,
+			protocol: "http",
+			host: "localhost",
+			mode: "local",
+			durableObjects: [],
+		};
+
+		await Promise.all([
+			registry.registerWorker(workerName, workerDefinition),
+			registry.unregisterWorker(workerName),
+		]);
+
+		// Verify the worker is no longer registered
+		const registeredWorkers = await registry.getRegisteredWorkers();
+		expect(registeredWorkers?.[workerName]).toBeUndefined();
+
+		// Verify that the worker file was removed
+		const workerFilePath = path.join(tempDir, workerName);
+		expect(fs.existsSync(workerFilePath)).toBe(false);
+	});
+
+	// TODO: We could keep track of workers registered by the current instance and unregister them on shutdown.
+	//       To maintain a single source of truth, the worker definitions could store a FilesystemWorkerRegistry "id".
+	it.skip("should ensure that there are no registered workers left if the registry shut down correctly", async () => {
+		const cb = vi.fn();
+		const cleanup = await registry.devRegistry(cb);
+
+		const workerName = "test-worker";
+		const workerDefinition: WorkerDefinition = {
+			port: 8787,
+			protocol: "http",
+			host: "localhost",
+			mode: "local",
+			durableObjects: [],
+		};
+
+		// Register the worker
+		await registry.registerWorker(workerName, workerDefinition);
+
+		// Shutdown the registry without providing the name of the worker
+		await cleanup();
+
+		// Verify the worker is no longer registered
+		const registeredWorkers = await registry.getRegisteredWorkers();
+		expect(registeredWorkers?.[workerName]).toBeUndefined();
+	});
+
+	// TODO: This is a difficult problem, but could potentially be solved by maintaining a registry that is
+	//       out-of-process of all wrangler instances. It could then guarantee (e.g. by using a mutex) that
+	//       there will be no read/write conflicts.
+	describe.concurrent("Concurrent read/write operations", () => {
```

Comment:
This is definitely a problem with the current implementation, but I'm not sure it's worth the effort to fix. In most realistic use cases there won't be multiple Wrangler instances with the same worker name running (outside of some of our tests), and even if we used a mutex the _user_-facing behaviour could still be unexpected (I don't think there's a reasonable intuitive outcome for running multiple wrangler dev sessions with the same name).